File: iosApp/iosApp/Feature/Business/Settings/BusinessSettingsScreen.swift
//
//  BusinessSettingsScreen.swift
//  iosApp
//
//  Created by BookkMe on 07.09.2025.
//  Copyright © 2025 BookkMe. All rights reserved.
//
import SwiftUI
import shared

struct BusinessSettingsScreen: View {
	
	@EnvironmentObject var navigationStack: NavigationStackHolder
	@StateViewModel var viewModel: BusinessSettingsViewModel
	
	init(id: shared.KotlinUuid) {
		self._viewModel = StateViewModel(wrappedValue: IOSBusinessDiKt.businessSettingsVM(id: id))
	}
	
	var body: some View {
		List {
			BusinessSettingsContent(viewModel: viewModel)
		}
		.toolbar {
			TextButton(viewModel.uiState.save) {
				viewModel.onSaveClick()
			}
		}
		.listSectionSpacing(.compact)
		.scrollDismissesKeyboard(.immediately)
		.withNavigationBar(viewModel.uiState.appBar)
		.handleNotifications(viewModel.uiState.notifications)
		.sendLifecycleEventsTo(viewModel)
		.handleNavigation(viewModel.uiState.navigation) { dest in
			switch dest {
			case is BusinessSettingsDestination.Back:
				navigationStack.popLast()
			default :
				break
			}
		}
	}
}

struct BusinessSettingsContent: View {
	
	let state: IOSBusinessSettingsState
	let viewModel: BusinessSettingsViewModel
	
	init(viewModel: BusinessSettingsViewModel) {
		self.state = IOSBusinessSettingsState.cast(viewModel.uiState)
		self.viewModel = viewModel
	}
	
	var body: some View {
		Section(BusinessRes.strings().business_settings_name_title.desc().localized()) {
			StateTextField(state.name) { text in
				viewModel.onNameChanged(name: text)
			}
			.textFieldStyle(.inList)
		}
		
		Section(BusinessRes.strings().business_settings_description_title.desc().localized()) {
			StateTextField(state.description_, textEditor: true) { text in
				viewModel.onDescriptionChanged(description: text)
			}
			.lineLimit(3, reservesSpace: true)
			.textFieldStyle(.inList)
		}
		
		let scheduleDays = IOSScheduleState.cast(state.schedule).list.items(DaySettingsState.self)
		Section(BusinessRes.strings().business_settings_schedule.desc().localized()) {
			ScheduleStrip(schedule: scheduleDays)
				.listRowInsets(EdgeInsets())
			if let expanded = scheduleDays.first(where: { $0.isVisible }) {
				ScheduleDay(state: expanded)
					.listRowInsets(EdgeInsets())
			}
		}
		.listRowSeparator(.hidden)
		
		Section(state.dayOffs.pickerTitle.localized()) {
			MultiPicker(
				state.dayOffs,
				pickerContent: {
					DateRangePicker(state: state.dateRange) {
						state.dayOffs.isPickerVisible = false
					}
				},
				itemContent: { item, onRemove in
					DayOffItem(item: item, onDeleteClick: onRemove)
				}
			)
		}
		
		Section {
			StateTextField(state.location) { text in
			}
			.textFieldStyle(.inList)
			
			TextButton(state.pickLocation, textAlignment: .leading) {
				viewModel.onPickLocationClicked()
			}
		} header : {
			Text(BusinessRes.strings().business_settings_location_title.desc().localized())
		} footer : {
			TextButton(state.testLocation, textAlignment: .leading) {
				viewModel.onTestLocationClick()
			}
		}
		
		Section(BusinessRes.strings().business_settings_address_title.desc().localized()) {
			StateTextField(state.address) { text in
				viewModel.onAddressChanged(address: text)
			}
			.textFieldStyle(.inList)
		}
		
		Section(BusinessRes.strings().business_settings_currency_title.desc().localized()) {
			PickerField(state.currency) { option in
				viewModel.onCurrencySelected(currencyUI: option as! CurrencyUI)
			}
			.textFieldStyle(.inList)
		}
		
		Section(BusinessRes.strings().business_settings_socials_title.desc().localized()) {
			StateTextField(state.instagram) { text in
				viewModel.onInstagramChanged(insta: text)
			}
			.textFieldStyle(.inList)
			
			StateTextField(state.telegram) { text in
				viewModel.onTelegramChanged(telegram: text)
			}
			.textFieldStyle(.inList)
			
			StateTextField(state.viber) { text in
				viewModel.onViberChanged(viber: text)
			}
			.textFieldStyle(.inList)
		}
	}
}

private struct DayOffItem: View {
    let item: PickerPresentation
    let onDeleteClick: () -> Void

    var body: some View {
		HStack(alignment: .center, spacing: 8) {
			Image(systemName: "calendar")
				.foregroundStyle(AppColors.primary)
			Text(item.displayName.localized())
				.font(.headline)
				.frame(maxWidth: .infinity, alignment: .leading)
			Button(action: onDeleteClick) {
				Image(systemName: "trash")
					.foregroundStyle(AppColors.error)
			}
			.buttonStyle(.plain)
		}
    }
}

private struct ScheduleStrip: View {

    let schedule: [DaySettingsState]

    var body: some View {
		HStack(spacing: 0) {
			ForEach(schedule, id: \.id) { day in
				DayOfWeekCell(state: day) {
					withAnimation {
						for d in schedule {
							d.isVisible = d.id == day.id ? !d.isVisible : false
						}
					}
				}
			}
		}
		.padding(10)
    }
}

private struct DayOfWeekCell: View {
    let state: any DaySettingsState
    let onClick: () -> Void

    var body: some View {
        Button(action: onClick) {
            Text(state.dayIndicator.localized())
                .font(.headline)
                .frame(width: 40, height: 40)
				.foregroundStyle(state.isActive.isChecked ? AppColors.onAction : AppColors.primary)
                .background(state.isActive.isChecked ? AppColors.buttonPrimary : Color.clear, in: Circle())
                .overlay(
                    Circle()
						.stroke(AppColors.primary, lineWidth: state.isVisible ? 2 : 0)
                )
        }
        .buttonStyle(.plain)
        .frame(maxWidth: .infinity)
    }
}

private struct ScheduleDay: View {
    let state: any DaySettingsState

    var body: some View {
		HStack(spacing: 8) {
			Text(state.dayIndicator.localized())
				.font(.headline)
				.frame(width: 40, height: 40)
				.foregroundStyle(AppColors.onAction)
				.background(state.isActive.isChecked ? AppColors.actionText : AppColors.inactive, in: Circle())
			Text(state.title.localized())
				.font(.headline)
				.frame(maxWidth: .infinity, alignment: .leading)
			StateSwitch(state: state.isActive) { newValue in
				state.isActive.onCheckedChange?(KotlinBoolean(bool: newValue))
			}
		}
		.padding(16)

		ForEach(state.intervals, id: \.id) { interval in
			TimeRow(timeSettings: interval) {
				withAnimation {
					state.onDeleteInterval(interval)
				}
			}
			.padding(.horizontal, 16)
			.padding(.bottom, 8)
		}

		TextButton(state.addTimeButton)
    }
}

private struct TimeRow: View {
    let timeSettings: any TimeSettingState
    let onDeleteClick: () -> Void

    var body: some View {
        HStack(spacing: 0) {
			TimePickerField(state: timeSettings.timeFromPicker)
				.textFieldStyle(.onElevated)
            Text("-")
                .font(.subheadline)
                .foregroundStyle(AppColors.secondary)
                .padding(.horizontal, 8)
            TimePickerField(state: timeSettings.timeToPicker)
				.textFieldStyle(.onElevated)
            Button(action: onDeleteClick) {
                Image(systemName: "xmark")
                    .foregroundStyle(AppColors.secondary)
            }
            .buttonStyle(.plain)
            .padding(.leading, 16)
        }.background(AppColors.elevated)
    }
}

#Preview {
	NavigationStack {
		BusinessSettingsScreen(id: shared.KotlinUuid.companion.random())
	}
}

File: feature/business/presentation/src/androidMain/kotlin/me/bookk/feature/business/presentation/settings/BusinessSettingsScreen.kt
package me.bookk.feature.business.presentation.settings

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Scaffold
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import dev.icerock.moko.resources.desc.desc
import me.bookk.android.feature.business.resources.BusinessRes
import me.bookk.designsystem.components.AppTopBar
import me.bookk.designsystem.components.Header
import me.bookk.designsystem.components.PickerField
import me.bookk.designsystem.components.TextButton
import me.bookk.designsystem.components.TextField
import me.bookk.designsystem.modifier.bottomShadow
import me.bookk.designsystem.theme.AppTheme
import me.bookk.designsystem.theme.ThemeMode
import me.bookk.designsystem.theme.color.LocalColors
import me.bookk.feature.business.presentation.settings.state.BusinessSettingsState

@Composable
internal fun BusinessSettingsScreen(state: BusinessSettingsState) {
    Scaffold(
        topBar = {
            AppTopBar(
                state = state.appBar,
                actions = {
                    TextButton(
                        state = state.save,
                        onClick = LocalBusinessSettingsEventListener.current.onSaveClick
                    )
                },
                onNavigationIconClick = LocalBusinessSettingsEventListener.current.onBackClick
            )
        },
        content = {
            Box(
                modifier = Modifier
                    .fillMaxSize()
                    .bottomShadow(32.dp, LocalColors.current.background)
            ) {
                Column(
                    modifier = Modifier
                        .padding(it)
                        .verticalScroll(rememberScrollState())
                        .padding(horizontal = 16.dp)
                        .padding(bottom = 24.dp),
                    verticalArrangement = Arrangement.spacedBy(4.dp)
                ) {
                    Column {
                        Header(BusinessRes.strings.business_settings_name_title.desc())
                        TextField(
                            state = state.name,
                            keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Text),
                            onValueChange = LocalBusinessSettingsEventListener.current.onNameChanged
                        )
                    }
                    Column {
                        Header(BusinessRes.strings.business_settings_description_title.desc())
                        TextField(
                            state = state.description,
                            keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Text),
                            onValueChange = LocalBusinessSettingsEventListener.current.onDescriptionChanged
                        )
                    }
                    Column {
                        Header(BusinessRes.strings.business_settings_location_title.desc())
                        Row(verticalAlignment = Alignment.CenterVertically) {
                            TextField(
                                modifier = Modifier.weight(1f),
                                state = state.location,
                                keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Text),
                                onValueChange = {}
                            )
                            TextButton(
                                state = state.pickLocation,
                                onClick = LocalBusinessSettingsEventListener.current.onPickLocationClick
                            )
                        }
                        TextButton(
                            state = state.testLocation,
                            onClick = LocalBusinessSettingsEventListener.current.onTestLocationClick
                        )
                    }
                    Column {
                        Header(BusinessRes.strings.business_settings_address_title.desc())
                        TextField(
                            state = state.address,
                            keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Text),
                            onValueChange = LocalBusinessSettingsEventListener.current.onAddressChanged
                        )
                    }
                    Column {
                        Header(BusinessRes.strings.business_settings_currency_title.desc())
                        PickerField(
                            state = state.currency,
                            onItemPicked = LocalBusinessSettingsEventListener.current.onCurrencySelected
                        )
                    }
                    Column {
                        Header(BusinessRes.strings.business_settings_socials_title.desc())
                        Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
                            TextField(
                                state = state.phone,
                                keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Phone),
                                onValueChange = LocalBusinessSettingsEventListener.current.onPhoneChanged
                            )
                            TextField(
                                state = state.telegram,
                                keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Text),
                                onValueChange = LocalBusinessSettingsEventListener.current.onTelegramChanged
                            )
                            TextField(
                                state = state.instagram,
                                keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Text),
                                onValueChange = LocalBusinessSettingsEventListener.current.onInstagramChanged
                            )
                            TextField(
                                state = state.viber,
                                keyboardOptions = KeyboardOptions(
                                    imeAction = ImeAction.Done,
                                    keyboardType = KeyboardType.Text
                                ),
                                onValueChange = LocalBusinessSettingsEventListener.current.onViberChanged
                            )
                        }
                    }
                }
            }
        }
    )
}

@Preview
@Composable
private fun PreviewDark() {
    AppTheme(themeMode = ThemeMode.DARK) {
        BusinessSettingsScreen(
            state = AndroidBusinessSettingsState(BusinessSettingsViewModel.createInitData())
        )
    }
}

@Preview
@Composable
private fun PreviewLight() {
    AppTheme(themeMode = ThemeMode.LIGHT) {
        BusinessSettingsScreen(
            state = AndroidBusinessSettingsState(BusinessSettingsViewModel.createInitData())
        )
    }
}